Expand documentation of custom random samplers.

tpapp opened the pull request:
Following [discussion](https://discourse.julialang.org/t/using-random-samplersimple/17864/3) on Discourse with @rfourquet, this PR expands the documentation for defining custom random samplers. Motivation is provided for the design, followed by examples that gradually introduce features and explain why/when they should be used.

No text was removed, but some parts were moved to other subsections.

Docstrings added for relevant methods and structures, and included in the docs.

The use of pre-existing types is emphasized.

tpapp committed:
Expand documentation of custom random samplers.

Following discussion on Discourse, this PR expands the documentation
for defining custom random samplers. Motivation is provided for the
design, followed by examples that gradually introduce features and
explain why/when they should be used.

No text was removed, but some parts were moved to other subsections.

Docstrings added for relevant methods and structures, and included in
the docs.

The use of pre-existing types is emphasized.

tpapp commented:
One thing that was unclear while I was doing this PR is whether to replace references to `eltype` by `Random.gentype`, eg in the docstring of `rand!` and the `Die` example. I can easily do this, if someone would clarify the preferred approach for me.

tpapp commented:
@rfourquet: friendly ping, could you please review at some point? I was hoping this would make it to 1.2.

StefanKarpinski commented:
We can definitely add docs in an RC.

rfourquet commented:
@tpapp I'm sorry this takes so long, the improvement here is amazing. I combined being on hollidays and sick, and couldn't find a moment to properly review. I should be able to manage by the end of the week.

tpapp commented:
@rfourquet: Thanks for getting back to me. No worries, take your time and get well soon!

rfourquet commented on stdlib/Random/docs/src/index.md:
```diff
@@ -146,38 +175,46 @@ julia> rand(Die(4), 3)
-In order to implement this decoupling for a custom type, a helper type can be used.
-Going back to our `Die` example: `rand(::Die)` uses random generation from a range, so
-there is an opportunity for this optimization:
+In order to implement this decoupling for a custom type, a custom helper type can be used. Going back to our `Die` example: `rand(::Die)` uses random generation from a range, so there is an opportunity for this optimization:
```
I think you should change "In order to implement this decoupling for a custom type", as you already showed how to do the decoupling for a custom type... maybe add "in cases where `SamplerSimple` is not suitable? (but the problem then is that we explain `SamplerSimple` was suitable in our `Die` example!)

tpapp replied:
I tried to clarify this, I hope the improved wording is clear.

rfourquet commented on stdlib/Random/docs/src/index.md:
```diff
@@ -228,8 +264,7 @@ Sampler(RNG::Type{<:AbstractRNG}, die::Die, ::Val{1}) = SamplerDie1(...)
 
-Of course, `rand` must also be defined on those types (i.e. `rand(::AbstractRNG, ::SamplerDie1)`
-and `rand(::AbstractRNG, ::SamplerDieMany)`).
+Of course, `rand` must also be defined on those types (i.e. `rand(::AbstractRNG, ::SamplerDie1)` and `rand(::AbstractRNG, ::SamplerDieMany)`). Alternatively, `SamplerTrivial` and `SamplerSimple` can be used if custom types are not necessary.
```
The "altenatively" seem to provide an alternative to defining `rand`. I would replace it by something like "Note that, as usual, ..."

rfourquet commented on stdlib/Random/src/Random.jl:
```diff
@@ -137,6 +137,23 @@ const Repetition = Union{Val{1},Val{Inf}}
+
+Return a sampler object that can be used to generate random values from `rng` for `x`.
+
+When `s = Sampler(rng, x, repetition)`, `rand(rng, s)` will be used to draw random values,
```
suggestion: `s` -> `sp` ?

rfourquet commented on stdlib/Random/src/Random.jl:
```diff
@@ -173,6 +202,15 @@ struct SamplerSimple{T,S,E} <: Sampler{E}
+
+Create a sampler that wraps the given value `x` and the `data`, with the type information
+from `Random.gentype(x)` (which should be defined). This is the default fall-back for
+values.
```
move "This is the default fall-back for values." to the `SamplerTrivial` docstring.

rfourquet reviewed (commented):
Excellent improvement, thank you!

rfourquet commented:
> One thing that was unclear while I was doing this PR is whether to replace references to eltype by Random.gentype

You can have a look at the discussion #27756 for background. Basically, people were not super motivated to introduce `gentype`, as it was seen mostly as a duplicate of `eltype`. So I kept it as an implementation detail till now. I'm still in favor to make it into the official API, but I think this should get a bit more support. A motivation is for some distributions which are not in `Random` (which defines only `rand` for types and collections), e.g. `Normal()`, or `Sampler`s: these objects are clearly not collections, such that defining `eltype` on them seems inadequate (hence the introduction of `gentype`). Your `DiscreteDistribution` is also such an example.

tpapp committed:
minor changes suggested

- clarify role of `gentype`

- use `sp` as an argument name for samplers

- corrections for docstrings

- clarify pedagogical motivation for custom samplers not using
  existing types

- various minor text corrections: spaces, emphases

tpapp commented on stdlib/Random/docs/src/index.md:
thanks, corrected

tpapp commented on stdlib/Random/docs/src/index.md:
thanks, clarified this

tpapp commented on stdlib/Random/docs/src/index.md:
```diff
@@ -126,12 +156,11 @@ julia> a = Vector{Die}(undef, 3); rand!(a)
+#### A simple sampler without pre-computed data
+
+Here we define a sampler for a collection. If no specialized algorithm or pre-computed data is required, it may be simplest to just return a random element, which can be implemented with a `SamplerTrivial` sampler, which is in fact the *default* fallback for values.
 
```
I am not sure what I was thinking when writing that. Made the other corrections and using emphasis consistently.

tpapp commented:
@rfourquet: Thanks for the review, I implemented the changes you requested.

Regarding

> introduce `gentype`, as it was seen mostly as a duplicate of `eltype`. So I kept it as an implementation detail till now. I'm still in favor to make it into the official API, but I think this should get a bit more support. 

I think that making `gentype` part of the API is the right decision, but I leave this for a future PR. Should I open an issue as a reminder?

tpapp commented:
Friendly bump: please let me know if I should do anything else to get this merged.

rfourquet commented on stdlib/Random/docs/src/index.md:
add a period or something at the end of this sentence

rfourquet commented on stdlib/Random/docs/src/index.md:
I'm confused here: you said 

> making gentype part of the API is the right decision, but I leave this for a future PR

with which I agree, but you document it now, making it public API...

If we keep a reference to `gentype` here in this document, we would need the approval of a core dev... and a rework of this sentence which is still not fully satisfying: the raison-d'être of `gentype` is to handle non-collections: collections have already `eltype` defined by definition, so saying `gentype` is used for computing element types for containers reinforces the idea that `gentype` is redundant. Moreover, the second sentence suggests that its fine to have `eltype != gentype`, which I don't think we should encourage... actually, I may even tend to believe this should not be allowed, i.e. either `eltype` or `gentype` should be defined, but not both... which is a point in favor of removing `gentype` altogether, while extending the scope of `eltype` (so that it's applicable beyond collections).

rfourquet commented on stdlib/Random/docs/src/index.md:
If we keep a reference to `gentype` here, I suggest something like:

"Note that  in this case, `Random.gentype` must *not* be defined (generated values are of type `T`).

rfourquet commented on stdlib/Random/docs/src/index.md:
```diff
@@ -194,10 +233,9 @@ Sampler(RNG::Type{<:AbstractRNG}, die::Die, r::Random.Repetition) =
+It's now possible to get a sampler with `sp = Sampler(rng, die)`, and use `sp` instead of `die` in any `rand` call involving `rng`. In the simplistic example above, `die` doesn't need to be stored in `SamplerDie` but this is often the case in practice.
 
-This pattern is so frequent that a helper type named `Random.SamplerSimple` is available,
+Of course, this pattern is so frequent that a helper type used above, named `Random.SamplerSimple`, is available,
```
```suggestion
Of course, this pattern is so frequent that the helper type used above, namely `Random.SamplerSimple`, is available,
```

rfourquet commented:
@tpapp thanks for the updates, I added a few more comments. But you seem to have missed few of my comments from the first batch? 

Concerning `gentype`: you offered to open an issue, and I think it's a good idea. You can also include content from a comment I just made inline in this PR, or I can add it to your issue.

tpapp committed:
Consistent ending dots, remove references to gentype.

rfourquet committed:
Update stdlib/Random/docs/src/index.md

Co-Authored-By: tpapp <tkpapp@gmail.com>

tpapp commented:
@rfourquet, thanks for the thorough second review. I addressed all your questions. In particular, I decided to remove all references to `Random.gentype` so that this can be merged.

In agreement with @StefanKarpinski, I would prefer to get this PR merged, with the understanding that it can always be improved later. Regardless of what happens to `gentype`, I think this is already an improvement.

rfourquet commented on stdlib/Random/docs/src/index.md:
````diff
@@ -198,8 +195,6 @@ end
 ```
 and that we *always* want to build an a alias table, regardless of the number of values needed (we learn how to customize this below). The methods
 ```julia
-Random.gentype(::Type{<:DiscreteDistribution}) = Int
````
Here it may be better to replace `gentype` by `eltype`.

tpapp replied:
Ok, adding back.

rfourquet commented:
Looks better for now with undocumenting `gentype`. But please check for the (very) few comments you didn't respond to, there is at least one which is important (unqualifed `probabilities` in `make_alias_table(probabilities)`)

tpapp committed:
define eltype for custom distribution, qualify probabilities

tpapp commented:
@rfourquet: I don't see a comment about the unqualified `probabilities`, but I corrected it; also added that `eltype` for the custom distribution.

tpapp committed:
remove all docstring references for gentype

This was making Documenter fail.

tpapp commented:
I fixed the docstring errors (by removing all remaining references to `Random.gentype`). The remaining CI error seems unrelated.

Unless there is anything else to address, it would be great to finally merge this.

rfourquet commented:
> I don't see a comment about the unqualified probabilities

It's there, along with 4 other comments. Check for "5 hidden conversations" when you reload the page. 

> Unless there is anything else to address, it would be great to finally merge this.

I didn't miss the fact that you wished this to be merged without unnecessary delay, I reviewed this PR at the pace I could. There is something left to address (cf. the "4 comments" mentioned above), but unfortunately it's already merged.

rfourquet commented:
Anyway, we can indeed fix that later, thanks for your contribution.

tpapp commented:
@rfourquet: thanks for pointing these out, I will fix them when I make a PR for #31968.

tpapp commented on stdlib/Random/docs/src/index.md:
```diff
@@ -126,12 +156,11 @@ julia> a = Vector{Die}(undef, 3); rand!(a)
-In order to define random generation out of objects of type `S`, the following method can be defined:
-`rand(rng::AbstractRNG, sp::Random.SamplerTrivial{S})`. Here, `sp` simply wraps an object of type `S`, which can be accessed via `sp[]`.
-Continuing the `Die` example, we want now to define `rand(d::Die)` to produce an `Int` corresponding to one of `d`'s sides:
+In order to define random generation out of objects of type `S`, the following method should be defined: `rand(rng::AbstractRNG, sp::Random.SamplerTrivial{S})`. Here, `sp` simply wraps an object of type `S`, which can be accessed via `sp[]`. Continuing the `Die` example, we want now to define `rand(d::Die)` to produce an `Int` corresponding to one of `d`'s sides:
```
No, this was accidental. Should I revert this in the next PR?

rfourquet replied:
No I think better to leave it like this now, unless you touch again this part. (I have a slight personal preference for split lines, as it's easier to read on my editor and makes it sometimes easier for git diffs and for commenting on github).

tpapp commented on stdlib/Random/docs/src/index.md:
```diff
@@ -146,38 +175,46 @@ julia> rand(Die(4), 3)
+#### An optimized sampler with pre-computed data
 
-#### Optimizing generation with cached computation between calls
+Consider a discrete distribution, where numbers `1:n` are drawn with given probabilities that some to one. When many values are needed from this distribution, the fastest method if using an [alias table](https://en.wikipedia.org/wiki/Alias_method). We don't provide the algorithm for building such a table here, but suppose it is available in `make_alias_table(probabilities)` instead, and `draw_num ...
```
Fixing these in the upcoming PR.

The alias table method is only needed for non-equal probabilities from a discrete/categorical distribution. It is used in `Distributions`. `Random` has no such sampler AFAIK.
